### ISSUE TITLE
big discrepancies between torch.jit.trace'd eval results an original model eval results

### ISSUE THREAD

**Issue by mikel-brostrom** [pytorch/pytorch#73353]
### 🐛 Describe the bug

I am exporting my model by:

```python
ts = torch.jit.trace(model, (im, im), strict=False)
optimize_for_mobile(ts)._save_for_lite_interpreter('/path/to/model.torchscript')
```

Then I load it in a LightningModule:

```python
class Model(pl.LightningModule):
    def __init__(self):
        super().__init__()

        # load my pretrained model
        self.model = torch.jit.load('/path/to/model.torchscript')

        self.f1 = F1Score(ignore_index=0, num_classes=2, average='macro', mdmc_average='samplewise')
        self.recall = Recall(ignore_index=0, num_classes=2, average='macro', mdmc_average='samplewise')
        self.precision_ = Precision(ignore_index=0, num_classes=2, average='macro', mdmc_average='samplewise')  # self.precision exists in 

    def forward(self, t0, t1):
        x = self.model(t0, t1)
        return x

    def validation_step(self, batch, batch_idx):

        t0_orig, t1_orig, t0, t1, mask_gt = batch
        mask_pred = self.forward(t0, t1)

        mask_pred = torch.where(mask_pred > 0.5, 1, 0)

        f1_score = self.f1(mask_pred, mask_gt)
        precision = self.precision_(mask_pred, mask_gt)
        recall = self.recall(mask_pred, mask_gt)

        stat_scores_ = stat_scores(mask_pred, mask_gt, num_classes=2, ignore_index=0, reduce='macro', mdmc_reduce='samplewise')
        # extract stat_scores for CHANGE class, ignore NO CHANGE
        stat_scores_ = stat_scores_[:,1,:]
        # reduce over batch
        stat_scores_ = torch.sum(stat_scores_, dim=0)
        tp = stat_scores_[0].to(torch.float)
        fp = stat_scores_[1].to(torch.float)
        tn = stat_scores_[2].to(torch.float)
        fn = stat_scores_[3].to(torch.float)

        # gpu tensor --> cpu numpy array
        mask_pred = mask_pred.detach().cpu().numpy()
        mask_pred = np.where(mask_pred > 0.5, 255, 0)

        metrics = {
            'val_f1_score': f1_score,
            'val_precision': precision,
            'val_recall': recall,
            'val_TP': tp,
            'val_FP': fp,
            'val_TN': tn,
            'val_FN': fn,
        }

        self.log_dict(metrics, on_epoch=True)
```

and finally I setup my dm and trainer:

```python
model = Model()
dm = DataModule(config)
trainer = pl.Trainer(
        accelerator='cpu',
    )
trainer.validate(model, datamodule=dm)
```

The original model is giving me:

```bash
DATALOADER:0 VALIDATE RESULTS
{'val_FN': 50533.0,
 'val_FP': 18351.80078125,
 'val_TN': 466346.1875,
 'val_TP': 120129.0,
 'val_f1_score': 0.7504137754440308,
 'val_loss': 0.23713822662830353,
 'val_precision': 0.8525729179382324,
 'val_recall': 0.6827183961868286}
```
The torch.jit.trace'd with optimization:

```bash
DATALOADER:0 VALIDATE RESULTS
{'val_FN': 78403.3984375,
 'val_FP': 411616.0,
 'val_TN': 73082.0,
 'val_TP': 92258.6015625,
 'val_f1_score': 0.26815277338027954,
 'val_precision': 0.18441727757453918,
 'val_recall': 0.5279602408409119}
```

The torch.jit.trace'd without optimization yields even worse results:

```bash
DATALOADER:0 VALIDATE RESULTS
{'val_FN': 107140.3984375,
 'val_FP': 442983.40625,
 'val_TN': 41714.6015625,
 'val_TP': 63521.6015625,
 'val_f1_score': 0.18113592267036438,
 'val_precision': 0.12397060543298721,
 'val_recall': 0.35545507073402405}
```

I am evaluating both models on the same machine and expect a small performance drop due to optimization but not that much. Is this a bug or am I missing something?

### Versions

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.15.5
Libc version: glibc-2.25

Python version: 3.6.9 (default, Dec  8 2021, 21:08:43)  [GCC 8.4.0] (64-bit runtime)
Nvidia driver version: 470.103.01
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

**Comment by xsacha** [comment]
What if you do:
```Diff
ts = torch.jit.trace(model, (im, im), strict=False)
- optimize_for_mobile(ts)._save_for_lite_interpreter('/path/to/model.torchscript')
+ ts.save('/path/to/model.torchscript')
```
instead?

**Comment by mikel-brostrom** [comment]
Hi @xsacha ,

> Also is (im, im) a similar shape to (t0, t1)?

(im, im) and (t0, t1)  have equal shapes

> Do they need to be different images?

Yes they need to be different

> What if you do:
Then I get what I wrote under `The torch.jit.trace'd without optimization yields even worse results:`

**Comment by mikel-brostrom** [comment]
I found out what the error was. The model was not set to evaluation mode by `model.eval()` before exporting. This means that layers such as bn are not deactivated and affects the evaluation. The evaluation mode was set automatically by the trainer for the original model but this is not possible for the traced model. Is there any way of warning of this @williamFalcon?
